specs/language/basic.tex

```diff
@@ -1,5 +1,97 @@
+
+\p There are three \textit{standard signed integer types}: \texttt{int16\_t},
+\texttt{int32\_t}, and \texttt{int64\_t}. Each of the signed integer types is
+explicitly named for the size in bits of the type's object representation. There
+is also the type alias \texttt{int} which is an alias of \texttt{int32\_t}. The
+standard signed integer types are collectively called \textit{signed integer
+types}.
+
+\p There are three \textit{standard unsigned integer types}: \texttt{uint16\_t},
+\texttt{uint32\_t}, and \texttt{uint64\_t}. Each of the unsigned integer types is
+explicitly named for the size in bits of the type's object representation. There
+is also the type alias \texttt{uint} which is an alias of \texttt{uint32\_t}. The
+standard unsigned integer types are collectively called \textit{unsigned integer
+types}\footnote{HLSL does not inherit C's default type rule, nor does it
+inherit the \texttt{unsigned} keyword.}.
+
+\p The signed integer types and unsigned integer types are collectively called
+\textit{integer types}. Integer types inherit the object representation of
+integers defined in \glsdesc{isoC23}\footnote{C23 adopts two's compliment as the
+object representation for integer types.}. Integer types shall satisfy the
+constraints defined in \glsdesc{isoCPP}, section \textbf{basic.fundamental}.
+
+\p There are three \textit{floating point types}: \texttt{half}, \texttt{float},
+and \texttt{double}. The \texttt{float} type is a 32-bit floating point type.
+The \texttt{double} type is a 64-bit floating point type. Both the
+\texttt{float} and \texttt{double} types have object representations as defined
+in \gls{IEEE754}. The \texttt{half} type may be either 16-bit or 32-bit as
+controlled by implementation defined compiler settings. If \texttt{half} is
+32-bit it will have an object representation as defined in \gls{IEEE754},
+otherwise it will have an object representation matching the \textbf{binary16}
+format defined in \gls{IEEE754}\footnote{IEEE-754 only defines a binary encoding
+for 16-bit floating point values, it does not fully specify the behavior of such
+types.}.
+
+\p Integer and floating point types are collectively called \textit{arithmetic
+types}.
+
+\p The \texttt{void} type is inherited from \gls{isoCPP}, which defines it as
+having an empty set of values and being an incomplete type that can never be
+completed. The \texttt{void} type is used to signify the return type of a
+function that returns no value. Any expression can be explicitly converted to
+\texttt{void}.
+
+\Sub{Special Classes}{Basic.types.special}
+
+\p A class type may be an \textit{ordinary class} or a \textit{special class}.
+An ordinary class is a class type that contains no members of special class,
+array of special class, or cv-qualified versions of these types. Special class
+objects can only exist in thread memory: they cannot be stored to any other
+memory space (\ref{Intro.Memory.Spaces}).
```

Comment:
This definition is recursive with no base case - it only defines that a class is a special class if it contains a special class, but doesn't explain why a class would be special otherwise.